main.py

```diff
@@ -1,68 +1,225 @@
+def ping(update: Update, context: CallbackContext):
+	dt=datetime.now(update.message.date.tzinfo)-update.message.date
+	update.message.reply_text(f'Ping is {dt.total_seconds():.2f}s')
+
+def escape_md(txt: str) -> str:
+	return escape_markdown(txt, 2)
+
+def get_mention(user: User):
+	return user.mention_markdown_v2()
+
+@message(Filters.status_update.new_chat_members)
+def new_chat_member(update: Update, context: CallbackContext):
+	handles=", ".join(get_mention(user) for user in update.message.new_chat_members)
+	update.message.reply_text(
+f"""{handles},
+いらっしゃいませ\\! \\[Welcome\\!\\]
+Welcome to this chat\\! Please read the rules\\.
+Добро пожаловать в чат\\! Прочти правила, пожалуйста\\.
+このチャットへようこそ！ ルールをお読みください。
+
+[rules](https://t\\.me/dev\\_meme/3667)""",
+parse_mode=ParseMode.MARKDOWN_V2)
+
+def is_admin(chat: Chat, user: User) -> bool:
+	# might wanna cache admins
+	status = chat.get_member(user.id).status
+	return status == 'creator' or status == 'administrator'
+
+def get_warn_target(message: Message) -> User | None:
+	'''
+	Returns the user that is supposed to be warned. It might be a bot.
+	Returns None if no warn target.
+	'''
+	if message.reply_to_message is not None and message.reply_to_message.from_user is not None:
+		return message.reply_to_message.from_user
+	return None
+
+class WarnPossibleResult(NamedTuple):
+	'''
+	can_warn -> can warn happen (user is admin, user is not a bot, warn target isnt bot)
+	reason -> reason why warn cannot happen, might be None
+	'''
+	is_possible: bool
+	reason: str | None
+
+def is_warn_possible(message: Message, command: str) -> WarnPossibleResult:
+	if message.from_user.is_bot: # might be redundant
+		return WarnPossibleResult(False, None)
+	if not is_admin(message.chat, message.from_user):
+		return WarnPossibleResult(False, f'You are not an admin')
+	target = get_warn_target(message)
+	if target is None:
+		return WarnPossibleResult(False, f'Please reply to a message with /{command}')
+	if target.is_bot:
+		return WarnPossibleResult(False, f'Bots cannot be warned')
+	return WarnPossibleResult(True, None)
+
+@command("warn")
+@filter_chat(private_chat_id, private_chat_username)
+def warn_member(update: Update, context: CallbackContext):
+	warn_pos = is_warn_possible(update.message, 'warn')
+	if not warn_pos.is_possible:
+		if warn_pos.reason is not None:
+			update.message.reply_text(warn_pos.reason, parse_mode=ParseMode.MARKDOWN_V2)
+		return
+	target = get_warn_target(update.message)
+	userid = target.id
+	warns = db.get_warns(userid) + 1
+	db.set_warns(userid, warns)
+	update.message.chat.send_message(
+		f'*{get_mention(target)}* recieved a warn\\! Now they have {warns} warns',
+		parse_mode=ParseMode.MARKDOWN_V2)
+
+@command("unwarn")
+@filter_chat(private_chat_id, private_chat_username)
+def unwarn_member(update: Update, context: CallbackContext):
+	can_warn = is_warn_possible(update.message, 'unwarn')
+	if not can_warn.is_possible:
+		if can_warn.reason is not None:

```

Comment:
pls refrain from using `is`/`is not` bc `not` is a keyword too and this could mean `x is (not None)` i.e. `x is True` to less experienced pythnon programmers - use `!=` instead for better readability pls (same thing on Lines 122, 151, 189)

Reply:
I saw in mypy that it's better to use "is not" for None checks, honestly I don't know the difference between this And != operator. 
`is not` works like !=, and I think it tries to mimic English language
I'll dIg into that later

Reply:
`is`/`is not` check for identity, `==`/`!=` for equality. `None`, `True`, `False` are all singletons, so it technically doesn't matter, but still. Readability.

Reply:
I remembered. It's better to use identity test, since some classes might break when they __eq__(None).
But still, we don't have such classes, so. Readability.